doc: Add an introduction and a design overview to the user guide (#674)
NomicFoundation/slang: documentation/public/user-guide/introduction.md

## What is Slang?

Slang is intended to be a modular Solidity compiler, specifically targeting code analysis and developer tooling. This means servicing tools with domain-specific APIs and, in general, facilitating working with and analyzing the Solidity source code. If you're in the editor writing Solidity or performing linting or additional validation, there's a chance that you are, or could be, running Slang!

## What Slang is not?

> First and foremost, it is not a replacement to `solc`, the standard Solidity compiler. We do not plan at the moment to support emitting EVM bytecode. Secondly, it is not meant to be used for formal verification of contracts or Solidity logic in general.

Reviewer: replacement to => replacement for

## Supporting multiple versions

> Over the years, the Solidity programming language has evolved quite a bit since its inception. Some features were introduced, some changed, while some eventually became obsolete and were removed altogether.

Reviewer: Remove "Over the years, "

> While it's good for a programming language to evolve and better serve the needs of its users, not being able to easily upgrade or re-deploy existing contracts poses a unique challenge. Older contracts that are still being used on the blockchain, written in older versions of Solidity, must remain understandable during the development process.

Reviewer 2: nit suggestion about "must remain understandable":

```suggestion
While it's good for a programming language to evolve and better serve the needs of its users, not being able to easily upgrade or re-deploy existing contracts poses a unique challenge. Developer tooling must be able to understand and consume older contracts, that are still being used on the blockchain, written in older versions of Solidity.
```

Author: Changed in https://github.com/NomicFoundation/slang/pull/674/commits/f8dcb7f8de77f2a941fdc73d78b05a9711250096

> 

Reviewer 2: Suggestion to add a paragraph at the end:

> This is why, instead of downloading a different Slang version for each Solidity version you want to use, you only need to use the latest Slang version, and initialize the Slang language API with the Solidity version instead. See [language specific guides](link) for more examples.

Author: Added, thanks! I've adapted the sentence to not repeat Solidity/Slang as much, hopefully it's still as clear.

> Because of that, Slang must be able to reason about different versions of Solidity - how the language grammar, name binding rules and semantics changed over multiple versions. One of our goals is to document differences as part of our [Solidity Specification](../solidity-specification/index.md).

Reviewer 2: ```suggestion
Because of that, Slang must be able to reason about different versions of Solidity; How the language grammar, name binding rules, and semantics have changed across different versions. One of our goals is to document differences as part of our [Solidity Specification](../solidity-specification/index.md).
```

Author: Changed in https://github.com/NomicFoundation/slang/pull/674/commits/f8dcb7f8de77f2a941fdc73d78b05a9711250096

## What is Slang?

Slang is intended to be a modular Solidity compiler, specifically targeting code analysis and developer tooling. This means servicing tools with domain-specific APIs and, in general, facilitating working with and analyzing the Solidity source code. If you're in the editor writing Solidity or performing linting or additional validation, there's a chance that you are, or could be, running Slang!

## What Slang is not?

> First and foremost, it is not a replacement for `solc`, the standard Solidity compiler. We do not plan at the moment to support emitting EVM bytecode. Secondly, it is not meant to be used for formal verification of contracts or Solidity logic in general.

Reviewer 2: discussion: distinction of production vs development codegen.

Author: Thanks; done in ae4b8969d6885a2acbcb16b221eadf6e7a885a69
